fix(backend): added setup structure to simplify adding new tests and removing duplicated code. Fixes #4630

NikeNano opened the pull request:
**Description of your changes:**
Addes setup functions for common parts of the tests to reduce duplicated code and simplify adding new test.  
Fixes #4630

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
<!--
   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.
-->
- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).
<!--
    **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update before release.
-->

NikeNano commented:
/assign @Bobgy

Bobgy commented:
Thank you! These refactorings help improve readability a lot!

Side note, the original issue included refactoring uploading pipelines to utility functions. Do you plan to work on that in this PR? #4630

A general comment is that, can we expand `w` and `b` variable names to proper words?

NikeNano commented:
> Side note, the original issue included refactoring uploading pipelines to utility functions. Do you plan to work on that in this PR? #4630

I misunderstood this and thought it was more a refactoring in general, I have now refactored the code to try to remove as much as possible of the uploading steps in to utility functions. 

> A general comment is that, can we expand `w` and `b` variable names to proper words when they are no longer local variables?

Changed this as well. 

As always thanks for the feedback.

NikeNano commented:
/retest

NikeNano commented:
/retest

Bobgy commented on backend/src/apiserver/server/pipeline_upload_server_test.go:
```diff
@@ -553,3 +447,43 @@ func TestDefaultUpdatedPipelineVersion(t *testing.T) {
+	return clientManager, resourceManager, server, bytesBuffer, writer, part
+}
+
+func uploadPipeline(method string, url string) (*resource.FakeClientManager, *resource.ResourceManager, *http.Request, *httptest.ResponseRecorder, PipelineUploadServer, *bytes.Buffer, *multipart.Writer, http.HandlerFunc) {
```
This is already improving existing tests a lot, so it's fine if you want to merge like this too.

Anyway, I'm giving a general comment if you want to improve it further:
It's better to write a utility function that does one thing well and can really encapsulate the action, because that makes it easy to **compose** multiple function calls to test any flexible things.
Right now, the `uploadPipeline` function also sets up the test, so we cannot call this method twice in one test case and we still have places that require a lot of boilerplate code for uploading the pipeline the second time.
Also, it returns so many variables making it really hard to understand what it is doing under the hood.

So I'd suggest creating a few `setup` methods that just do the setup and then the `uploadPipeline` method should take required inputs from the setup method, so tests can setup once and upload multiple times with different parameters.

Bobgy replied:
I didn't look through the complex set up, so I guess you got your difficulties achieving this, feel free to discuss that with me.

NikeNano replied:
Thanks for the feedback, I will break it up and fix it :)

NikeNano commented:
/hold

NikeNano commented:
please take a look again @Bobgy, I have broke it up in to smaller reusable functions now.

Bobgy commented on backend/src/apiserver/server/pipeline_upload_server_test.go:
```diff
@@ -534,22 +414,59 @@ func TestDefaultUpdatedPipelineVersion(t *testing.T) {
+	return server
+}
+
+func uploadPipeline(server PipelineUploadServer, method string, url string, body io.Reader, requestKey string,
```
Hi @NikeNano, thank you for the update!

I apologize, I just realized we could first iterate on `uploadPipeline` interface. When we finalize on the interface, you'll be able to finish the implementation in one go without back and forths with huge cost on your side.

Some parameters seems unnecessary:
Method will always be "POST", requestKey is always "Content-Type", requestValue always come from the same variable.

I think the conceptual way to look at whether this interface is a good fit is
 first thinking about what parameters `uploadPipeline(...)` needs. e.g. I think it needs the `uploadServer`, `pipelineName?`, `pipelineContent`, that's pretty much it. Then you can figure out what types they are.

After that, you can think about `uploadPipelineVersion(...)`, it needs `uploadServer`, `pipelineId`, `pipelineContent`, `pipelineVersionName`.

I understand some of the tests are testing an error case that you need special content in the request:
```
writer.CreateFormFile("uploadfile", "hello-world.yaml")
writer.Close()
```

So it might be a little hard to achieve the ideal interface, I totally understand that.

I think you've done a wonderful refactoring that greatly improved readability of these test cases.
Just get rid of the duplicative parameters and we are good to go.

NikeNano replied:
Thank you, appreciate all the feedback! I might look in to the other tests later and try to refactor them as well if needed.

NikeNano replied:
I update and removed the unnecessary parameters you pointed out from the uploadPipeline function.

NikeNano commented:
ping @Bobgy

NikeNano commented:
/unhold

Bobgy commented on backend/src/apiserver/server/pipeline_upload_server_test.go:
```diff
@@ -534,22 +415,58 @@ func TestDefaultUpdatedPipelineVersion(t *testing.T) {
+	return server
+}
+
+func uploadPipeline(server PipelineUploadServer, url string, body io.Reader, writer *multipart.Writer, uploadFunc func(http.ResponseWriter, *http.Request)) *httptest.ResponseRecorder {
```
Sorry for the delay, I've been busy for a while.

Just realized one last thing, the `server` parameter doesn't seem to be used in the function body. Shall we remove it?

NikeNano replied:
Yeah, you are right. I should have double checked more carefully. Fixed.

Bobgy reviewed (commented):
/lgtm

Bobgy commented:
/lgtm
/approve
Thank you @NikeNano !
